src/Plutarch/SafeMoney.hs

```diff
@@ -178,10 +186,101 @@ pdiscreteValue' (Tagged (AssetClass (cs, tn))) =
             PDiscrete t <- pmatchC p
             pure $ psingletonValue # pconstant cs # pconstant tn # (pextract # t)
 
-{-
-pdiscreteValue' (Tagged (AssetClass (cs, tn))) = phoistAcyclic $
-  plam $ \p ->
-    psingletonValue
-      # pconstant cs
-      # pconstant tn
-      # Tagged.puntag p -}
+-- | @since 1.0.0
+type PRateDecimal = PFixedDecimal 1000000
+
+-- | @since 1.0.0
+newtype PExchangeRate (from :: k) (to :: k) (s :: S)
+    = PExchangeRate (Term s (PTagged '(from, to) PRateDecimal))
+
+deriveGeneric ''PExchangeRate
+
+-- | @since 1.0.0
+deriving via
+    (DerivePNewtype (PExchangeRate from to) (PTagged '(from, to) PRateDecimal))
+    instance
+        PlutusType (PExchangeRate from to)
+
+-- | @since 1.0.0
+deriving via
+    (DerivePNewtype (PExchangeRate from to) (PTagged '(from, to) PRateDecimal))
+    instance
+        PIsData (PExchangeRate from to)
+
+-- | @since 1.0.0
+deriving via
+    (DerivePNewtype (PExchangeRate from to) (PTagged '(from, to) PRateDecimal))
+    instance
+        PEq (PExchangeRate from to)
+
+-- | @since 1.0.0
+deriving via
+    (DerivePNewtype (PExchangeRate from to) (PTagged '(from, to) PRateDecimal))
+    instance
+        POrd (PExchangeRate from to)
+
+-- | @since 1.0.0
+deriving anyclass instance PShow (PExchangeRate from to)
+
+-- | @since 1.0.0
+deriving via
+    (DerivePNewtype (PExchangeRate from to) (PTagged '(from, to) PRateDecimal))
+    instance
+        PTryFrom a PRateDecimal => PTryFrom a (PExchangeRate from to)
+
+-- | @since 1.0.0
+instance PTryFrom PData (PAsData (PExchangeRate from to)) where
+    type
+        PTryFromExcess PData (PAsData (PExchangeRate from to)) =
+            PTryFromExcess PData (PAsData PRateDecimal)
+    ptryFrom' d k = ptryFrom' @_ @(PAsData PRateDecimal) d $ k . first punsafeCoerce
+
+{- | Make @PExchangeRation@ with given @PIntegers. Two Integers are nominator and denominator.
+
+ @since 1.0.0
+-}
+pexchangeRatio ::
+    forall k.
+    forall (to :: k) (from :: k) (s :: S).
+    Term s (PInteger :--> PInteger :--> PExchangeRate from to)
+pexchangeRatio = phoistAcyclic $ plam $ \n d -> pexchangeRatio' n d
```

Comment:
This will explode if the denominator is zero. We should probably have some better error messaging in such a case if we're willing to tolerate the partiality; if not, working inside `PMaybe` is not the worst idea.

Reply:
I think using `PMaybe` would be a good idea since we have good helpers like `pfromJust`